### DISCUSSION TITLE
诸如 wp_loop 这样命名的变量，前缀 wp 是什么意思？

### DISCUSSION THREAD

**Discussion by Gwill** [cpp-main/cpp-tbox#36] [Q&A] [answered]
诸如 wp_loop 这样命名的变量，前缀 wp 是什么意思？

**Comment by baldwin-law** [comment, marked as answer]
> 诸如 wp_loop 这样命名的变量，前缀 wp 是什么意思？

为了区分成员变量的所有权，如果不是此对象持有的成员，仅仅是引用，则用wp_loop表示，如过改对象持有loop的整个生命周期，则用sp_loop表示。 （从智能指针中得到启发。）

**Comment by hevake** [comment]
cpp-tbox早期是有采用智能指针的，以sp_表示shared_ptr，wp_表示weak_ptr。
shared_ptr 是有持有对象的生命期的，而 weak_ptr 则没有。
后面发现智能指针具有一定的“传染性”（如果库用了智能指针，则使用者就会被迫使用智能指针）让使用者不“自由”，则弃用了智能指针。
但 sp_ 与 wp_ 的传统被保留了下来。它们用于区别是持有所指对象，还是仅仅是引用了所指对象。
如果是持有，那么在父对像析构的时候一定要一起释放它。如果是引用，则可以不理会。
再后来，这个习惯也逐渐被淡化，没有再刻意使用 sp_ 与 wp_ 进行标识了。
